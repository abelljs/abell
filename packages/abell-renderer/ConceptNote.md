## Single File Components

```jsx
<AbellComponent name="Description">
  <p>{{ props.children }}</p>
</AbellComponent>

<html>
<body>
  <h2>Hello, World</h2>
  <Description>My description text</Description>
</body>
</html>
```

---

## Compile to JS?

```jsx
{{
  function Description(props) {
    return `
      <p>${props.children}</p>
    `
  }
}}

<html>
<body>
  <h2>Hello, World</h2>
  {{ Description({children: 'My description text'}) }}
</body>
</html>
```

---

## Build on top of Vite? YES!!

---

## Parsing Component Logic

- parsing component
- executing component code


Executing Component
1. Component Call
```jsx
{{
  const Hello = abellRequire('./component/Hello.abell');
}}
// <Hello props={xyz: 'hi'}>
{{ Hello({xyz: 'hi'}).html }}
```

2. Component Defination
```jsx
<AbellComponent>
  <template><div>Hello {{ props.xyz }}</div></template>
</AbellComponent>
```

3. Component Defination in JS
```js
function abellRequire(abellComponentPath) {
  const abellComponentContent = fs.readFileSync(abellComponentPath, 'utf8');

  // Hello(props).html
  return (props) => {
    const compiledCode = compile(abellComponentContent, {props});
    const template = compiledCode.match(/<template>(.*?)</template>/);
    return {
      html: template
    } 
  };
}
```


```jsx
<AbellComponent>
  <template>
    <div>Hello</div>
  </template>
</AbellComponent>
```

```js
const abellComponentContent = fs.readFileSync(abellComponentPath, 'utf8')

function Hello(props) {
  const compiledCode = compile(abellComponentContent, {props});
  const template = compiledCode.match(/<template>(.*?)</template>/);
  return {
    html: template
  }
}

module.exports = Hello;
```

```js
const Hello = requireAbellComponent(abellComponentPath)
```

```js
function requireAbellComponent(abellComponentPath) {
  const abellComponentContent = fs.readFileSync(abellComponentPath, 'utf8');

  return function AbellComponent(props) {

  }
}
```


How can I get components array before executing component function?

Do I even need this?

How else can I bundle?


A:
We cannot let people use `props` in styles and scripts. It will end up creating different bundles for different calls.