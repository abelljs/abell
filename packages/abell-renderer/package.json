{
  "name": "abell-renderer",
  "version": "1.0.0-alpha.7",
  "description": "",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "test": "jest",
    "dev": "yarn example:dev",
    "example:dev": "yarn ae dev",
    "dev:with-ts-run": "ts-run src/example.ts -w",
    "start": "ts-run ./src/example.ts",
    "serve": "yarn build && node dist/example.js",
    "build": "tsc",
    "eslint": "eslint",
    "prepublishOnly": "yarn eslint && yarn build"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "acorn": "^8.3.0",
    "htmlparser2": "^6.1.0",
    "postcss-selector-parser": "^6.0.6",
    "stylis": "^4.0.10"
  },
  "devDependencies": {
    "@types/stylis": "^4.0.1"
  }
}
