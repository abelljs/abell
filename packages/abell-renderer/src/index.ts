import fs from 'fs';
import path from 'path';

import {
  UserOptions,
  OutputWithComponent,
  UserOptionsAllowComponents,
  UserOptionsBase
} from './types';
import { getAbellInBuiltSandbox } from './utils/general-utils';
import { compile } from './compiler';

const defaultUserOptions: UserOptions = {
  basePath: '',
  dangerouslyAllowRequire: false,
  allowComponents: false,
  filename: '<sandbox>.abell'
};

export function render(
  abellTemplate: string,
  userSandbox: Record<string, unknown>,
  options: UserOptionsAllowComponents
): OutputWithComponent;

export function render(
  abellTemplate: string,
  userSandbox?: Record<string, unknown>,
  options?: UserOptions
): string;

export function render(
  abellTemplate: string,
  // eslint-disable-next-line @typescript-eslint/ban-types
  userSandbox: Record<string, unknown> = {},
  options: UserOptions = defaultUserOptions
): string | OutputWithComponent {
  // Set initial variables
  options.basePath =
    options.basePath ||
    (options.filename && path.dirname(options.filename)) ||
    '';

  const { builtInFunctions, subComponents } = getAbellInBuiltSandbox(options);

  const templateSandbox = {
    ...builtInFunctions,
    ...userSandbox
  };

  const html = compile(abellTemplate, templateSandbox, options);

  if (options.allowComponents) {
    return {
      html,
      components: subComponents
    };
  }

  return html;
}

export function engine(
  { dangerouslyAllowRequire } = { dangerouslyAllowRequire: false }
) {
  return (
    filePath: string,
    options: UserOptionsBase,
    callback: (_: null, rendered: string) => string
  ): string => {
    // define the template engine
    const content = fs.readFileSync(filePath, 'utf-8');
    const rendered = render(content, options, {
      basePath: path.dirname(filePath),
      dangerouslyAllowRequire
    });
    return callback(null, rendered);
  };
}

export default {
  render,
  engine
};
