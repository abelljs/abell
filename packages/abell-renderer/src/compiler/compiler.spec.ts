import { compile } from './index';

describe('compile()', () => {
  test('should successfully compile single expressions', () => {
    const out = compile('<body>{{ 3 + 4 }}</body>');
    expect(out).toMatchSnapshot();
  });

  test('should successfully compile with declarations', () => {
    const abellCode = `
    {{
      import fs from 'fs';
      import path from 'path';
      import { compile } from 'abell-renderer';
    }}
    <body>
      {{ 3 + 4 }}
      <b>{{ 'Helloo'.toUpperCase() }}</b>
      <div>
        {{
          [1, 2, 3].map(i => 
            i * 2
          )
        }}
      </div>
    </body>
    `;
    const out = compile(abellCode);
    expect(out).toMatchSnapshot();
  });
});
