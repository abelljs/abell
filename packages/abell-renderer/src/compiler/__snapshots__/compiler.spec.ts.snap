// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile() should successfully compile single expressions 1`] = `
"
  
  const e = (val) => val ? val : '';
  export const html = (props) => {
    const Abell = { props };
    const console = { 
      log: (...args) => { console.log('[<anonymous>.abell]:', ...args); return ''; } 
    };
    return \`<body>\${e( 3 + 4 )}</body>\`
  };
  export default html;
  "
`;

exports[`compile() should successfully compile with declarations 1`] = `
"
  
      import fs from 'fs';
      import path from 'path';
      import { compile } from 'abell-renderer';
    
  const e = (val) => val ? val : '';
  export const html = (props) => {
    const Abell = { props };
    const console = { 
      log: (...args) => { console.log('[<anonymous>.abell]:', ...args); return ''; } 
    };
    return \`<body>
      \${e( 3 + 4 )}
      <b>\${e( 'Helloo'.toUpperCase() )}</b>
      <div>
        \${e(
          [1, 2, 3].map(i => 
            i * 2
          )
        )}
      </div>
    </body>\`
  };
  export default html;
  "
`;
